source/date_time.c

```diff
@@ -0,0 +1,826 @@
+        case 't':
+            next = time_string[++index];
+            switch (next) {
+                case 'H':
+                case 'h':
+                    next = time_string[++index];
+                    switch (next) {
+                        case 'U':
+                        case 'u':
+                            return 4;
+                        default:
+                            return -1;
+                    }
+                case 'U':
+                case 'u':
+                    next = time_string[++index];
+                    switch (next) {
+                        case 'E':
+                        case 'e':
+                            return 2;
+                        default:
+                            return -1;
+                    }
+                default:
+                    return -1;
+            }
+        case 'M':
+        case 'm':
+            next = time_string[++index];
+            switch (next) {
+                case 'O':
+                case 'o':
+                    next = time_string[++index];
+                    switch (next) {
+                        case 'N':
+                        case 'n':
+                            return 1;
+                        default:
+                            return -1;
+                    }
+                default:
+                    return -1;
+            }
+        case 'W':
+        case 'w':
+            next = time_string[++index];
+            switch (next) {
+                case 'E':
+                case 'e':
+                    next = time_string[++index];
+                    switch (next) {
+                        case 'D':
+                        case 'd':
+                            return 3;
+                        default:
+                            return -1;
+                    }
+                default:
+                    return -1;
+            }
+        case 'F':
+        case 'f':
+            next = time_string[++index];
+            switch (next) {
+                case 'R':
+                case 'r':
+                    next = time_string[++index];
+                    switch (next) {
+                        case 'I':
+                        case 'i':
+                            return 5;
+                        default:
+                            return -1;
+                    }
+                default:
+                    return -1;
+            }
+        default:
+            return -1;
+    }
+}
+
+/* Get the 0-11 monthy number from a string representing Month. Case insensitive and will stop on abbreviation*/
+static int get_month_number_from_str(const char *time_string, size_t startIndex, size_t stopIndex) {
+    if (stopIndex - startIndex < 3) {
+        return -1;
+    }
+
+    size_t index = startIndex;
+
+    char c = time_string[index];
+    char next = 0;
+
+    /* it's ugly but this should compile down to EXACTLY 3 comparisons and no memory allocations */
+    switch (c) {
```

Comment:
See above suggestions - don't write triply-nested switches by hand.

If you don't want to use code generators, and still want a single switch, consider:

```
#define STR_TRIPLET_TO_INDEX(str) \
  (   ((uint32_t)(uint8_t)tolower((str)[0]) << 0) | \
      ((uint32_t)(uint8_t)tolower((str)[1]) << 8) | \
      ((uint32_t)(uint8_t)tolower((str)[2]) << 16) )

// ...
if (!time_string[0] || !time_string[1] || !time_string[2]) return -1;

switch(STR_TRIPLET_TO_INDEX(str)) {
  case STR_TRIPLET_TO_INDEX("jan"): return 0;
  case STR_TRIPLET_TO_INDEX("feb"): return 1;
// ...
  default: return -1;
}
```

Note that I'm not sure if STR_TRIPLET_TO_INDEX will work in the case labels, but you should be able to precompute it.

Reply:
will try it out.

Reply:
Another option for this would be something like:

```
uint32_t tmp = 0;
strncpy(&tmp, str, 3);

// This AND forces the string to lowercase (assuming ASCII)
switch (aws_ntohl(tmp) & ~0x20202020) {
  // ...
}
```

This might be a bit faster, as it avoids having all those individual tolower calls. If you can guarantee that str is at least 4 bytes long then a memcpy (and mask of the last byte) would be even better.